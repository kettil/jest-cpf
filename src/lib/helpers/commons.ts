import { ExtendPromiseType } from '../types';

export const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms, ms));

export const extendPromise = <T>(promise: Promise<T>): ExtendPromiseType<T | Error> => {
  let isPending = true;

  return {
    isPending: () => isPending,
    promise: new Promise<T | Error>((resolve) => {
      promise.then(
        (v: T) => {
          isPending = false;
          resolve(v);
        },
        (error: Error) => {
          isPending = false;
          resolve(error);
        },
      );
    }),
  };
};
