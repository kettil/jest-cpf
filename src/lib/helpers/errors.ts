export class TestError extends Error {
  readonly data: string;

  constructor(message?: string, data = '') {
    super(message);

    this.data = data;
  }
}
