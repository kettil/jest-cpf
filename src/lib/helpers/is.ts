import { CoverageSummaryType } from '../types';

export const isObject = (data: unknown): data is Record<string, unknown> => typeof data === 'object' && data !== null;

export const isSummary = (data: unknown): data is CoverageSummaryType =>
  isObject(data) && typeof data.pct === 'number' && typeof data.total === 'number' && typeof data.covered === 'number';

export const isGlobalThreshold = (threshold: unknown): threshold is { global: Record<string, unknown> } =>
  isObject(threshold) && isObject(threshold.global);
