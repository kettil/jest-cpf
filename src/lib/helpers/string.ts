import { join, dirname, basename } from 'path';

export const cleanFileFromRootPath = (file: string, rootDirectory: string) =>
  file.slice(Math.max(0, rootDirectory.length + 1));

export const cleanFileFromExtensions = (file: string) => {
  const path = dirname(file);
  const base = basename(file);
  const name = base.slice(0, Math.max(0, base.indexOf('.') + 1));

  return join(path, name);
};

export const mapPerLine = (value: string, callback: (line: string) => string) =>
  value
    .split('\n')
    .map(callback)
    .join('\n');

export const padFill = (value: string, length: number, fill: string) =>
  [...new Array(Math.max(0, length - value.length))].fill(fill).join('');

export const padLeft = (value: string, length: number, fill: string): string => padFill(value, length, fill) + value;

export const padRight = (value: string, length: number, fill: string): string => value + padFill(value, length, fill);

export const formatPercent = (value: number): string => {
  const [pre, post = ''] = value.toString().split('.');

  return [padLeft(pre, 3, ' '), padRight(post, 2, '0')].join('.');
};
