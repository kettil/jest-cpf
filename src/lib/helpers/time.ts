import { DateTime } from 'luxon';

export const startTime = () => DateTime.utc();

export const getRunTime = (time: DateTime) => {
  const runtime = DateTime.utc().diff(time, ['minutes', 'seconds']);

  return runtime.toFormat('mm:ss');
};
