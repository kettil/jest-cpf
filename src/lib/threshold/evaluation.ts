import { ThresholdColorType, ThresholdGroupType, ThresholdType, ThresholdValueType } from '../types';
import { thresholdKeys } from './settings';

export const isSuccess = (percent: number, lines: number, limit: number) =>
  (limit < 0 && -1 * lines >= limit) || (limit >= 0 && percent >= limit);

export const evaluationFileStatus = (value: ThresholdValueType, limit: number): ThresholdColorType => {
  if (typeof value === 'object') {
    const percent = value.percent;
    const lines = value.uncovered;

    switch (true) {
      case isSuccess(percent, lines, limit):
        return 'success';

      default:
        return 'error';
    }
  }

  return 'error';
};

export const evaluationFile = (
  threshold: ThresholdGroupType,
  thresholdValue: ThresholdType<number>,
): ThresholdColorType => {
  for (const key of thresholdKeys) {
    const value = threshold.threshold[key];
    const limit = thresholdValue[key];
    const status = evaluationFileStatus(value, limit);

    if (status === 'error') {
      return status;
    }
  }

  return 'success';
};

export const evaluationFiles = (thresholds: ThresholdGroupType[], thresholdValue: ThresholdType<number>) => {
  const data: Record<ThresholdColorType, number> = { success: 0, error: 0 };

  for (const threshold of thresholds) {
    data[evaluationFile(threshold, thresholdValue)] += 1;
  }

  return data;
};
