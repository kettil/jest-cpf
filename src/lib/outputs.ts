/* eslint-disable no-console */
import { SingleBar } from 'cli-progress';
import { table } from 'table';
import { TestError } from './helpers/errors';
import { padLeft, mapPerLine } from './helpers/string';
import styling from './helpers/styling';
import createTableData from './threshold/createTableData';
import { tableConfig } from './threshold/settings';
import { ThresholdColorType, ThresholdGroupType, BarReturnType, ThresholdType } from './types';

export const outputError = (error: unknown) => {
  if (error instanceof Error) {
    if (error instanceof TestError) {
      console.error(styling('|> An error was thrown during testing!', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(styling('|> Output', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(mapPerLine(error.data, (line) => styling('|> ', 'red') + line));
    } else {
      console.error(mapPerLine(error.stack ?? `Error: ${error.message}`, (line) => styling('|> ', 'red') + line));
    }
  } else {
    console.error(error);
  }

  console.log('');
};

export const outputMetrics = (status: Record<ThresholdColorType, number>, runtime: string) => {
  const types = Object.keys(status) as ThresholdColorType[];
  const maxChars = types.reduce((p, c) => Math.max(p, status[c]), 0).toString().length;

  console.log(styling('Summary', 'underline'));
  console.log(`├─ Success: ${styling(padLeft(status.success.toString(), maxChars, ' '), 'green')}`);
  console.log(`└─ Error:   ${styling(padLeft(status.error.toString(), maxChars, ' '), 'red')}`);
  console.log('');

  console.log(`Runtime: ${runtime} (mm:ss)`);
  console.log('');
};

export const outputProgress = (count: number): BarReturnType => {
  const bar = new SingleBar({
    format: `${styling('{bar}', 'red')} {percentage}% | {value}/{total} Files`,
    barCompleteChar: '\u2588',
    barIncompleteChar: '\u2591',
    hideCursor: true,
  });

  bar.start(count, 0);

  return {
    increment: () => bar.increment(),
    stop: () => {
      bar.stop();
      console.log('');
    },
  };
};

export const outputStatus = (countFiles: number, countTestFiles: number, cpuCount: number, cpuUsed: number) => {
  console.log('');
  console.log(`CPU detected: ${styling(cpuCount, 'red')} (${styling(cpuUsed, 'red')} are used)`);
  console.log(`Files found:  ${styling(countFiles, 'red')} (${styling(countTestFiles, 'red')} with tests)`);
  console.log('');
};

export const outputTable = (thresholds: ThresholdGroupType[], thresholdLimits: ThresholdType<number>) => {
  const thresholdRows = createTableData(thresholds, thresholdLimits);

  console.log(table(thresholdRows, tableConfig));
};
