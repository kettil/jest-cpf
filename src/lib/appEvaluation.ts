import { DateTime } from 'luxon';
import appExecute from './appExecute';
import { cleanUpByFiles } from './files/cleanUpFiles';
import { getRunTime } from './helpers/time';
import { outputTable, outputMetrics } from './outputs';
import { evaluationFiles } from './threshold/evaluation';
import { thresholdForFilesWithoutTests } from './threshold/settings';
import { UnpackedPromise } from './types';

const appEvaluation = (time: DateTime, data: UnpackedPromise<ReturnType<typeof appExecute>>) => {
  const { cleanedUpFiles, thresholdData, thresholdLimits } = data;

  // remove all files with tests
  cleanUpByFiles(
    cleanedUpFiles,
    thresholdData.map(({ file }) => file),
  ).forEach((file) => {
    // add files without tests
    thresholdData.push({ file, threshold: thresholdForFilesWithoutTests });
  });

  const status = evaluationFiles(thresholdData, thresholdLimits);

  outputTable(thresholdData, thresholdLimits);
  outputMetrics(status, getRunTime(time));

  return status;
};

export default appEvaluation;
