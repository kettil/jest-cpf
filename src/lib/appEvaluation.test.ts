import { DateTime } from 'luxon';
import appEvaluation from './appEvaluation';
import { getRunTime } from './helpers/time';
import { outputTable, outputMetrics } from './outputs';
import { evaluationFiles } from './threshold/evaluation';

jest.mock('./outputs');
jest.mock('./threshold/evaluation');
jest.mock('./helpers/time');

const evaluationFilesMock = evaluationFiles as jest.Mock;
const outputMetricsMock = outputMetrics as jest.Mock;
const outputTableMock = outputTable as jest.Mock;
const getRunTimeMock = getRunTime as jest.Mock;

describe('appEvaluation()', () => {
  test('it should work', () => {
    const time = (jest.fn() as unknown) as DateTime;
    const data = {
      thresholdLimits: { branches: 95, functions: 95, lines: 95, statements: 95 },
      cleanedUpFiles: ['/path/to/project/src/index.ts', '/path/to/project/src/app.ts'],
      thresholdData: [
        {
          file: '/path/to/project/src/index.ts',
          threshold: {
            branches: { percent: 90, uncovered: 3 },
            functions: { percent: 90, uncovered: 3 },
            lines: { percent: 100, uncovered: 3 },
            statements: { percent: 90, uncovered: 3 },
          },
        },
      ],
    };

    evaluationFilesMock.mockReturnValueOnce({ error: 6, success: 8 });
    getRunTimeMock.mockReturnValueOnce('13:42');

    const status = appEvaluation(time, data);

    expect(status).toEqual({ error: 6, success: 8 });

    expect(evaluationFilesMock).toHaveBeenCalledTimes(1);
    expect(evaluationFilesMock).toMatchSnapshot('evaluationFiles');

    expect(outputMetricsMock).toHaveBeenCalledTimes(1);
    expect(outputMetricsMock).toMatchSnapshot('outputMetrics');

    expect(outputTableMock).toHaveBeenCalledTimes(1);
    expect(outputTableMock).toMatchSnapshot('outputTable');

    expect(getRunTimeMock).toHaveBeenCalledTimes(1);
  });
});
