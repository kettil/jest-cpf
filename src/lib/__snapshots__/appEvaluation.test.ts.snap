// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appEvaluation() it should work: evaluationFiles 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "file": "/path/to/project/src/index.ts",
          "threshold": Object {
            "branches": Object {
              "percent": 90,
              "uncovered": 3,
            },
            "functions": Object {
              "percent": 90,
              "uncovered": 3,
            },
            "lines": Object {
              "percent": 100,
              "uncovered": 3,
            },
            "statements": Object {
              "percent": 90,
              "uncovered": 3,
            },
          },
        },
        Object {
          "file": "/path/to/project/src/app.ts",
          "threshold": Object {
            "branches": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "functions": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "lines": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "statements": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
          },
        },
      ],
      Object {
        "branches": 95,
        "functions": 95,
        "lines": 95,
        "statements": 95,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "error": 6,
        "success": 8,
      },
    },
  ],
}
`;

exports[`appEvaluation() it should work: outputMetrics 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "error": 6,
        "success": 8,
      },
      "13:42",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`appEvaluation() it should work: outputTable 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "file": "/path/to/project/src/index.ts",
          "threshold": Object {
            "branches": Object {
              "percent": 90,
              "uncovered": 3,
            },
            "functions": Object {
              "percent": 90,
              "uncovered": 3,
            },
            "lines": Object {
              "percent": 100,
              "uncovered": 3,
            },
            "statements": Object {
              "percent": 90,
              "uncovered": 3,
            },
          },
        },
        Object {
          "file": "/path/to/project/src/app.ts",
          "threshold": Object {
            "branches": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "functions": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "lines": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
            "statements": Object {
              "percent": 0,
              "uncovered": Infinity,
            },
          },
        },
      ],
      Object {
        "branches": 95,
        "functions": 95,
        "lines": 95,
        "statements": 95,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
